File: frontend/src/lib/services/largeOrders.ts
/**
 * Large Orders (Layer 3 block) service.
 * One net large-order bubble per trading day for a symbol over a date range,
 * for plotting on the 1D candlestick chart (Pine volume-cluster style).
 */

import { apiGet } from '../api';

export interface LargeOrderDay {
  date: string;        // ICT trading date YYYY-MM-DD
  time: number;        // unix seconds, UTC-midnight of the VN date (daily candle x)
  side: number;        // net side: 1=BUY if net_delta>=0 else 2=SELL
  net_delta: number;   // buy notional - sell notional (signed)
  buy_value: number;
  sell_value: number;
  total_value: number; // buy + sell notional (drives size tier)
  buy_qty: number;
  sell_qty: number;
  total_qty: number;   // buy + sell shares (volume label)
  num_trades: number;
  block_count: number;
}

export interface LargeOrdersResponse {
  symbol: string;
  blocks: LargeOrderDay[];
}

export interface GetLargeOrdersParams {
  fromDate?: string;   // YYYY-MM-DD
  toDate?: string;     // YYYY-MM-DD
  minValue?: number;   // only blocks with notional >= this
}

export const fetchLargeOrders = async (
  symbol: string,
  params: GetLargeOrdersParams = {},
): Promise<LargeOrdersResponse> => {
  const q = new URLSearchParams({ symbol });
  if (params.fromDate) q.set('from_date', params.fromDate);
  if (params.toDate) q.set('to_date', params.toDate);
  if (params.minValue != null) q.set('min_value', String(params.minValue));
  return apiGet<LargeOrdersResponse>(`/large-orders?${q.toString()}`);
};

File: backend/app/api/v1/routes/large_orders.py
"""Large-order (Layer 3 block) API endpoints."""
from datetime import date, timedelta
from typing import Optional

from fastapi import APIRouter, Depends, Query
from clickhouse_connect.driver import Client

from app.db.clickhouse import get_clickhouse_client
from app.services.large_orders_service import (
    LargeOrdersService,
    LargeOrdersResponse,
)

router = APIRouter()


def get_large_orders_service(
    clickhouse_client: Client = Depends(get_clickhouse_client),
) -> LargeOrdersService:
    return LargeOrdersService(clickhouse_client)


@router.get("/large-orders", response_model=LargeOrdersResponse)
async def get_large_orders(
    symbol: str = Query(..., description="Symbol, e.g. FPT"),
    from_date: Optional[str] = Query(
        None, description="Start date YYYY-MM-DD (default: 400 days ago)"
    ),
    to_date: Optional[str] = Query(
        None, description="End date YYYY-MM-DD (default: today)"
    ),
    min_value: Optional[float] = Query(
        None, ge=0, description="Only blocks with notional >= this value"
    ),
    service: LargeOrdersService = Depends(get_large_orders_service),
):
    """One net large-order bubble per trading day for a symbol over a range.

    Per day: net delta = buy notional - sell notional (sign = bubble side),
    `total_value` drives the size tier, `total_qty` is the volume label, and
    `time` aligns to the day's 1D candle.
    """
    to_day = date.fromisoformat(to_date) if to_date else date.today()
    from_day = date.fromisoformat(from_date) if from_date else to_day - timedelta(days=400)
    return service.get_blocks(
        symbol=symbol.strip().upper(),
        from_day=from_day,
        to_day=to_day,
        min_value=min_value,
    )
